tests/android/MASVS-CODE/MASTG-TEST-0x27-1.md

```diff
@@ -0,0 +1,44 @@
+
+## Steps
+
+1. Examine the application's code (see @MASTG-TECH-0023)
+2. Look for occurrences of WebViews being used and examine if they are configured with a custom `WebViewClient`.
+3. Search for and inspect the following interception callback functions for the `WebViewClient`:
+
+- `shouldOverrideUrlLoading` allows your application to either abort loading pages with suspicious content by returning `true` or allow the WebView to load the URL by returning `false`. Considerations:
+    - This method is not called for POST requests.
+    - This method is not called for XmlHttpRequests, iFrames, "src" attributes included in HTML or `<script>` tags. Instead, `shouldInterceptRequest` should take care of this.
+- `shouldInterceptRequest` allows the application to return the data from resource requests. If the return value is null, the WebView will continue to load the resource as usual. Otherwise, the data returned by the `shouldInterceptRequest` method is used. Considerations:
+    - This callback is invoked for a variety of URL schemes (e.g., `http(s):`, `data:`, `file:`, etc.), not only those schemes which send requests over the network.
+    - This is not called for `javascript:` or `blob:` URLs, or for assets accessed via `file:///android_asset/` or `file:///android_res/` URLs.
+  In the case of redirects, this is only called for the initial resource URL, not any subsequent redirect URLs.
+    - When Safe Browsing is enabled, these URLs still undergo Safe Browsing checks but the developer can allow the URL with `setSafeBrowsingWhitelist` or even ignore the warning via the `onSafeBrowsingHit` callback. Safe Browsing can also fully be disabled by using `setSafeBrowsingEnabled(false)`.
+
+As you can see there are a lot of points to consider when testing the security of WebViews that have a WebViewClient configured, so be sure to carefully read and understand all of them by checking the [`WebViewClient` Documentation](https://developer.android.com/reference/android/webkit/WebViewClient "WebViewClient").
+
+## Observation
+
+The output could contain references to `WebViewClient` or calls to `shouldInterceptRequest`, `shouldOverrideUrlLoading` and `setSafeBrowsingEnabled`.
+
+## Evaluation
+
+The test case fails if the `WebView` has a custom `WebViewClient` and one of the following is true:
+
+- SafeSearch is disabled via `setSafeBrowsingEnabled(false)`
+- The `WebViewClient` is missing the `shouldOverrideUrlLoading` or `shouldInterceptRequest` handlers
+- The `shouldOverrideUrlLoading` or `shouldInterceptRequest` handlers do not correctly prevent untrusted data from being loaded in the `WebView`
+
+If the `WebView` does not have a custom `WebViewClient`, then any navigation event will automatically trigger the default browser.
```

Comment:
Not sure the last sentence is needed, since the test case is only for webviews with custom clients?